samcli/lib/build/workflow_config.py

```diff
@@ -69,12 +69,21 @@
     manifest_name="go.mod",
     executable_search_paths=None)
 
+PROVIDED_MAKE_CONFIG = CONFIG(
+    language="provided",
+    dependency_manager=None,
+    application_framework=None,
+    manifest_name="Makefile",
+    executable_search_paths=None)
+
 
 class UnsupportedRuntimeException(Exception):
     pass
 
+class UnsupportedBuilderException(Exception):
+    pass
 
-def get_workflow_config(runtime, code_dir, project_dir):
+def get_workflow_config(runtime, code_dir, project_dir, metadata=None):
```

Comment:
Also if you decide to use `BuildMethod` anyway, I don't think it is good idea to pass metadata all the way to this level. We have `build_function` method that does translate `codeuri` -> `code_dir` which `get_workflow_config` method uses. Now if you pass raw metadata, then get_workflow_config now have the logic to do that transformation to runtime. Now this cleaner function has convoluted logic of choosing between two runtimes. I would suggest two approaches:
1. Do this conversion in build_function method and pass the correct runtime in this method based on your logic.
2. Abstract out these conversions to Function class so that it is responsible to provide a local build_runtime based on all parameters.

I like approach one better because it is easier to do, and keeps all build related logic at one place. Probably abstract out whole thing in a _runtime_resolver method.

Reply:
yeah moving selector process to its own method was what I had in mind.